Code/Framework/AzCore/AzCore/Memory/PoolAllocator.cpp

```diff
@@ -665,8 +674,7 @@ namespace AZ
     // [9/15/2009]
     //=========================================================================
     PoolSchema::PoolSchema()
-        : m_impl(new (AZStd::stateless_allocator().allocate(sizeof(PoolSchemaImpl), AZStd::alignment_of<PoolSchemaImpl>::value))
-                 PoolSchemaImpl())
+        : m_impl(nullptr)
```

Comment:
The allocators need to fully initialize themselves inside their constructors. The reason for this is to ensure that the destruction order is correct. For instance, say that the `PoolSchemaImpl` object internally used some other dependent allocator, like the `SystemAllocator`. In order to ensure that the `SystemAllocator` is still valid when the `PoolSchema`'s destructor is run, the `SystemAllocator`'s constructor has to complete before the `PoolSchema` constructor completes. The way that is guaranteed is by having each allocator construct all of their dependent allocators. This is why the `SystemAllocator` calls `AllocatorInstance<OSAllocator>::Get()` in its constructor:
https://github.com/o3de/o3de/blob/69dbcd08a56539315bfb0472984daf0f46e7a966/Code/Framework/AzCore/AzCore/Memory/SystemAllocator.cpp#L33

Reply:
The SystemAllocator is fully constructed before the PoolAllocatorHelper's constructor completes, but not before the PoolSchema's constructor completes. Is that sufficient? The order with my change is
```
TestPoolAllocator()->PoolAllocatorHelper()->SimpleSchemaAllocator()->AllocatorBase()
                                |                                   |
                                |                                   -->SimpleSchemaAllocator::Create()->PoolSchema()
                                -->PoolSchema::Create()->stateless_allocator::allocate()->AZ_OS_MALLOC
                                                                     ->PoolSchemaImpl()->AllocatorInstance<SystemAllocator>::Get()
                                                                                                    ->PoolAllocation<PoolSchemaImpl>()->SystemAllocator::allocate()->HphaSchema::allocate()
```
The reason I moved the PoolsSchemaImpl construction is because the SimpleSchemaAllocator's constructor always creates a PoolSchema inside SimpleSchemaAllocator::Create(). There are no parameters for pageSize, minAllocationSize, or maxAllocationSize there, so there is no way for me to forward that info to PoolSchemaImpl's constructor. By creating PoolSchemImple inside PoolSchema::Create(), I have the option to have the PoolAllocatorHelper constructor take those arguments and use them to create the appropriate Impl.

I can think of a couple of alternatives to my implementation, I'd like to know which you think is best:
1. Keep what I have here, if it is sufficient to have SystemAllocator constructed after PoolSchema() completes, but before PoolAllocatorHelper() completes.
2. Would it be sufficient to simply call AllocatorInstance<SystemAllocator>::Get() in PoolSchema(), that way it is constructed before PoolSchema() completes?
3. I could add an alternative constructor and Create function to SimpleSchemaAllocator with pageSize, minAllocation, and maxAllocation parameters. Then have the PoolAllocatorHelper(pageSize, minAllocation, maxAllocation) constructor use that one instead. The only reason I didn't do this initially was that modifying the SimpleSchemaAllocator seemed like it was getting outside the purview of modifying the PoolAllocator. But now that I understand things better it seems like a reasonable solution to me, although it's bleeding implementation details of the underlying schema into the SimpleSchemaAllocator.

Reply:
Thanks for the diagram. It looks like this will work as written. I was thrown off by the fact that this is a `Schema` instead of an `Allocator`. Because `Schema::Create` is called in the constructor of the `Allocator` type, we still have the guarantee that the dependent allocators' constructors complete before the `PoolAllocator`'s does.